### PR TITLE
fix(runtime): initialize TransactionState when creating runtime instance

### PR DESCRIPTION
**Pull request by jimjbrettj**
## Changes

- Set transaction state as part of runtime context
- Make it so that the expected type of a KeyOwnershipProof is a scale encoded option for both babe and grandpa
See [babe docs](https://spec.polkadot.network/#sect-babeapi_generate_key_ownership_proof) and [grandpa docs](https://spec.polkadot.network/#sect-grandpaapi_generate_key_ownership_proof)

Note: The reason this change didn't work before is because we do not have proper nil checks on parameter values when creating a runtime instance. Perhaps this would be a good thing to discuss and add to avoid issues like this happening in the future.

## Tests

```sh
go test ./lib/babe --tags=integration 
go test ./lib/grandpa --tags=integration 
go test ./lib/runtime/wasmer
```

## Primary Reviewer

@kishansagathiya

### CODE REVIEW

File: lib/runtime/wasmer/exports.go [hunk 1]
```diff
@@ -110,13 +113,18 @@ func (in *Instance) BabeGenerateKeyOwnershipProof(slot uint64, authorityID [32]b
...
 
-	keyOwnershipProof := types.OpaqueKeyOwnershipProof{}
+	var keyOwnershipProof *types.OpaqueKeyOwnershipProof
 	err = scale.Unmarshal(encodedKeyOwnershipProof, &keyOwnershipProof)
 	if err != nil {
 		return nil, fmt.Errorf("scale decoding key ownership proof: %w", err)
 	}
 
-	return keyOwnershipProof, nil
+	// TODO check with team if returning error is preferred or should we keep just returning empty value
```
Comment (EclesioMeloJunior) [inline lib/runtime/wasmer/exports.go:122]: I tend to agree with returning an error since we expect that key to report an equivocation

**Review by EclesioMeloJunior — APPROVED** [review]

File: lib/runtime/wasmer/interfaces.go [hunk 2]
```diff
@@ -58,6 +60,7 @@ type BasicNetwork interface {
 // TransactionState is the interface for the transaction state.
 type TransactionState interface {
 	AddToPool(vt *transaction.ValidTransaction) common.Hash
+	PopWithTimer(timerCh <-chan time.Time) (transaction *transaction.ValidTransaction)
```
Comment (qdm12) [inline lib/runtime/wasmer/interfaces.go:63]: Why is this addition needed? It compiles fine with it commented 🤔

File: lib/runtime/interfaces.go [hunk 3]
```diff
@@ -50,4 +52,5 @@ type BasicStorage interface {
 // TransactionState interface for adding transactions to pool
 type TransactionState interface {
 	AddToPool(vt *transaction.ValidTransaction) common.Hash
+	PopWithTimer(timerCh <-chan time.Time) (transaction *transaction.ValidTransaction)
```
Comment (qdm12) [inline lib/runtime/interfaces.go:55]: Why is this addition needed? It compiles fine with it commented 🤔

File: lib/babe/verify_integration_test.go [hunk 4]
```diff
@@ -481,13 +461,13 @@ func TestVerifyAuthorshipRight_Equivocation(t *testing.T) {
...
 
-	err = verifier.verifyAuthorshipRight(&block.Header)
+	err = verificationManager.VerifyBlock(&block.Header)
 	require.NoError(t, err)
 
 	err = babeService.blockState.AddBlock(block2)
 	require.NoError(t, err)
 
-	err = verifier.verifyAuthorshipRight(&block2.Header)
+	err = verificationManager.VerifyBlock(&block2.Header)
```
Comment (kishansagathiya) [inline lib/babe/verify_integration_test.go:470]: This test is meant to test `verifyAuthorshipRight`, we can't call `VerifyBlock` directly. We must be calling `verifyAuthorshipRight`

File: lib/runtime/wasmer/exports.go [hunk 5]
```diff
@@ -110,13 +113,18 @@ func (in *Instance) BabeGenerateKeyOwnershipProof(slot uint64, authorityID [32]b
 		return nil, fmt.Errorf("executing %s: %w", runtime.BabeAPIGenerateKeyOwnershipProof, err)
 	}
 
-	keyOwnershipProof := types.OpaqueKeyOwnershipProof{}
+	var keyOwnershipProof *types.OpaqueKeyOwnershipProof
 	err = scale.Unmarshal(encodedKeyOwnershipProof, &keyOwnershipProof)
```
Comment (kishansagathiya) [inline lib/runtime/wasmer/exports.go:117]:
```suggestion
	err = scale.Unmarshal(encodedKeyOwnershipProof, keyOwnershipProof)
```

keyOwnershipProof is already a pointer. `&keyOwnershipProof` here becomes a pointer to pointer.

File: lib/runtime/wasmer/exports.go (hunk 1 above)
Comment (kishansagathiya) [reply lib/runtime/wasmer/exports.go:122]: I disagree. All these functions are only limited to just make a runtime call and return results as go structs.
If we get a nil key ownership proof, that still means that we called runtime function correctly and converted the go struct right way. This runtime func returns key ownership proof as scale encoded option. So, `keyOwnershipProof` could be nil.

I don't mind using `errEmptyKeyOwnershipProof`, but this is not the place where we use it.

File: lib/runtime/wasmer/exports.go [hunk 6]
```diff
@@ -297,13 +306,17 @@ func (in *Instance) GrandpaGenerateKeyOwnershipProof(authSetID uint64, authority
 		return nil, err
 	}
 
-	keyOwnershipProof := types.GrandpaOpaqueKeyOwnershipProof{}
+	var keyOwnershipProof *types.GrandpaOpaqueKeyOwnershipProof
 	err = scale.Unmarshal(encodedOpaqueKeyOwnershipProof, &keyOwnershipProof)
```
Comment (kishansagathiya) [inline lib/runtime/wasmer/exports.go:310]:
```suggestion
	err = scale.Unmarshal(encodedOpaqueKeyOwnershipProof, keyOwnershipProof)
```

**Review by kishansagathiya — CHANGES_REQUESTED** [review]
```
        keyOwnershipProof := types.GrandpaOpaqueKeyOwnershipProof{}
	err = scale.Unmarshal(encodedOpaqueKeyOwnershipProof, &keyOwnershipProof)
```

if `encodedOpaqueKeyOwnershipProof` contains the value scaled encoded nil, would keyOwnershipProof be set to nil?

File: lib/runtime/wasmer/exports.go (hunk 1 above)
Comment (jimjbrettj) [reply lib/runtime/wasmer/exports.go:122]: I think this is a good point, I will update

File: lib/runtime/wasmer/exports.go [hunk 7]
```diff
@@ -110,13 +113,18 @@ func (in *Instance) BabeGenerateKeyOwnershipProof(slot uint64, authorityID [32]b
...
 	err = scale.Unmarshal(encodedKeyOwnershipProof, &keyOwnershipProof)
 	if err != nil {
 		return nil, fmt.Errorf("scale decoding key ownership proof: %w", err)
 	}
 
-	return keyOwnershipProof, nil
+	// TODO check with team if returning error is preferred or should we keep just returning empty value
+	if keyOwnershipProof == nil {
+		return types.OpaqueKeyOwnershipProof{}, fmt.Errorf("%w", errEmptyKeyOwnershipProof)
+	}
```
[the remark this answers is not in the document]
Comment (jimjbrettj) [reply lib/runtime/wasmer/exports.go:125]: Saw @kishansagathiya asked this too so answering here. Yes, because the expected type is an option and Gossamer handles options as pointers, where the 0 case is nil

File: lib/babe/verify_integration_test.go (hunk 4 above)
Comment (jimjbrettj) [reply lib/babe/verify_integration_test.go:470]: So with how we use the VerificationManager in babe, the verifies is a private object so cannot retrieve it to call this receiver which is why I call VerifyBlock. Should I export or create a getter in order to call this specific function? Open to ideas on how to do this

File: lib/runtime/wasmer/exports.go (hunk 5 above)
Comment (jimjbrettj) [reply lib/runtime/wasmer/exports.go:117]: Yes that is the intended behavior now. Scale takes a pointer to the object, which in this case is a pointer, so a pointer to a pointer is intentional.

File: lib/runtime/wasmer/exports.go (hunk 6 above)
Comment (jimjbrettj) [reply lib/runtime/wasmer/exports.go:310]: Yes that is the intended behavior now. Scale takes a pointer to the object, which in this case is a pointer, so a pointer to a pointer is intentional.

File: lib/runtime/wasmer/exports.go [hunk 8]
```diff
@@ -297,13 +306,17 @@ func (in *Instance) GrandpaGenerateKeyOwnershipProof(authSetID uint64, authority
...
-	keyOwnershipProof := types.GrandpaOpaqueKeyOwnershipProof{}
+	var keyOwnershipProof *types.GrandpaOpaqueKeyOwnershipProof
 	err = scale.Unmarshal(encodedOpaqueKeyOwnershipProof, &keyOwnershipProof)
 	if err != nil {
 		return nil, fmt.Errorf("scale decoding: %w", err)
 	}
 
-	return keyOwnershipProof, nil
+	if keyOwnershipProof == nil {
+		return types.GrandpaOpaqueKeyOwnershipProof{}, fmt.Errorf("%w", errEmptyKeyOwnershipProof)
```
[the remark this answers is not in the document]
Comment (jimjbrettj) [reply lib/runtime/wasmer/exports.go:316]: Ended up removing this error message

File: lib/runtime/wasmer/exports.go [hunk 9]
```diff
@@ -110,13 +113,18 @@ func (in *Instance) BabeGenerateKeyOwnershipProof(slot uint64, authorityID [32]b
...
+	var keyOwnershipProof *types.OpaqueKeyOwnershipProof
 	err = scale.Unmarshal(encodedKeyOwnershipProof, &keyOwnershipProof)
 	if err != nil {
 		return nil, fmt.Errorf("scale decoding key ownership proof: %w", err)
 	}
 
-	return keyOwnershipProof, nil
+	// TODO check with team if returning error is preferred or should we keep just returning empty value
+	if keyOwnershipProof == nil {
+		return types.OpaqueKeyOwnershipProof{}, fmt.Errorf("%w", errEmptyKeyOwnershipProof)
```
[the remark this answers is not in the document]
Comment (jimjbrettj) [reply lib/runtime/wasmer/exports.go:124]: Ended up removing this error message

File: lib/runtime/wasmer/interfaces.go (hunk 2 above)
Comment (jimjbrettj) [reply lib/runtime/wasmer/interfaces.go:63]: good catch, not sure why that ended up being added, thank ya

File: lib/runtime/wasmer/exports.go [hunk 10]
```diff
@@ -110,13 +110,18 @@ func (in *Instance) BabeGenerateKeyOwnershipProof(slot uint64, authorityID [32]b
...
 
-	keyOwnershipProof := types.OpaqueKeyOwnershipProof{}
+	var keyOwnershipProof *types.OpaqueKeyOwnershipProof
 	err = scale.Unmarshal(encodedKeyOwnershipProof, &keyOwnershipProof)
 	if err != nil {
 		return nil, fmt.Errorf("scale decoding key ownership proof: %w", err)
 	}
 
-	return keyOwnershipProof, nil
+	// TODO check with team if returning error is preferred or should we keep just returning empty value
```
Comment (EclesioMeloJunior) [inline lib/runtime/wasmer/exports.go:119]:
```suggestion
```

**Review by EclesioMeloJunior — APPROVED** [review]
LGTM! Just a small adjustment

File: lib/grandpa/vote_message.go [hunk 11]
```diff
@@ -284,6 +287,8 @@ func (s *Service) reportEquivocation(stage Subround, existingVote *SignedVote, c
 	opaqueKeyOwnershipProof, err := runtime.GrandpaGenerateKeyOwnershipProof(setID, pubKey)
 	if err != nil {
 		return fmt.Errorf("getting key ownership proof: %w", err)
+	} else if opaqueKeyOwnershipProof == nil {
+		return fmt.Errorf("%w", errEmptyKeyOwnershipProof)
```
Comment (kishansagathiya) [inline lib/grandpa/vote_message.go:291]: why this error wrapping?

File: lib/runtime/wasmer/exports_test.go [hunk 12]
```diff
@@ -1103,10 +1104,10 @@ func TestInstance_GrandpaGenerateKeyOwnershipProof(t *testing.T) {
 	identityPubKey, _ := ed25519.NewPublicKey(identity)
 	authorityID := identityPubKey.AsBytes()
 
-	encodedOpaqueKeyOwnershipProof, err := instance.GrandpaGenerateKeyOwnershipProof(uint64(0), authorityID)
-	require.NoError(t, err)
+	res, err := instance.GrandpaGenerateKeyOwnershipProof(uint64(0), authorityID)
 	// Since the input is not valid with respect to the instance, an empty proof is returned
-	require.Empty(t, encodedOpaqueKeyOwnershipProof)
+	require.NoError(t, err)
+	require.Nil(t, res)
```
Comment (kishansagathiya) [inline lib/runtime/wasmer/exports_test.go:1110]: is this renaming and moving around necessary?

**Review by kishansagathiya — APPROVED** [review]
have small comments, but approving to avoid one more review-iteration.

File: lib/babe/verify.go [hunk 13]
```diff
@@ -360,6 +363,8 @@ func (b *verifier) submitAndReportEquivocation(
 	keyOwnershipProof, err := runtimeInstance.BabeGenerateKeyOwnershipProof(slot, offenderPublicKey)
 	if err != nil {
 		return fmt.Errorf("getting key ownership proof from runtime: %w", err)
+	} else if keyOwnershipProof == nil {
+		return fmt.Errorf("%w", errEmptyKeyOwnershipProof)
```
[the remark this answers is not in the document]
Comment (jimjbrettj) [reply lib/babe/verify.go:367]: I thought it would be good to do, but upon further review, I do not think it is needed. Will update in both cases.

File: lib/runtime/wasmer/exports_test.go (hunk 12 above)
Comment (jimjbrettj) [reply lib/runtime/wasmer/exports_test.go:1110]: In this case I actually did think it was helpful, since to me it is misleading to call the variable encoded since it is unmarshalled. Since there is already a diff from adding the nil require check, I thought it was okay to rename. 

However res is not that descriptive so I will change to opaqueKeyOwnershipProof
